Tryouts/Plugins/ApplicationPlugins/process explorer/backend/ModuleProcessMonitor/Processes/ProcessInfoGeneratorWindows.cs

```diff
@@ -0,0 +1,363 @@
+// 
+//      http://www.apache.org/licenses/LICENSE-2.0.
+// 
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership. Unless required by applicable law or agreed
+// to in writing, software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions
+// and limitations under the License.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Management;
+using System.Runtime.InteropServices;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using ModuleProcessMonitor.Logging;
+
+namespace ModuleProcessMonitor.Processes;
+
+[System.Runtime.Versioning.SupportedOSPlatform("windows")]
+internal class ProcessInfoGeneratorWindows : ProcessInfoManager
+{
+    private readonly ILogger<ProcessInfoManager> _logger;
+    private readonly ConcurrentDictionary<int, PerformanceCounter> _cpuPerformanceCounters;
+    private readonly ConcurrentDictionary<int, PerformanceCounter> _memoryPerformanceCounters;
+    private ManagementEventWatcher? _watcher;
+    private readonly object _lock = new();
+    public ProcessInfoGeneratorWindows(ILogger<ProcessInfoManager>? logger)
+    {
+        _logger = logger ?? NullLogger<ProcessInfoGeneratorWindows>.Instance;
+        _cpuPerformanceCounters = new();
+        _memoryPerformanceCounters = new();
+    }
+
+    public ProcessInfoGeneratorWindows(EventHandler<int> sendNewProcess,
+        EventHandler<int> sendTerminatedProcess, EventHandler<int> sendModifiedProcess,
+        ILogger<ProcessInfoManager>? logger = null)
+        : this(logger)
+    {
+        base.SetEvents(sendModifiedProcess, sendNewProcess, sendTerminatedProcess);
+    }
+
+    public override int? GetParentId(Process? process)
+    {
+        if (process == null)
+        {
+            return null;
+        }
+
+        int ppid = 0;
+
+        using ManagementObject mo =
+            new ManagementObject(
+                $"win32_process.handle='{process.Id}'");
+        try
+        {
+            mo.Get();
+            ppid = Convert.ToInt32(mo["ParentProcessId"]);
+        }
+        catch (Exception exception)
+        {
+            if (process.Id > 0)
+            {
+                _logger.ManagementObjectPpid(process.Id, exception);
+            }
+        }
+
+        return ppid;
+
+    }
+
+    public override float GetMemoryUsage(Process process)
+    {
+        int memsize;
+
+        if (!_memoryPerformanceCounters.TryGetValue(
+                process.Id,
+                out var perf))
+        {
+            perf = new();
+            perf.CategoryName = "Process";
+            perf.CounterName = "Working Set - Private";
+            perf.InstanceName = process.ProcessName;
+            _memoryPerformanceCounters[process.Id] = perf;
```

Comment:
This way the code is still not thread-safe. It won't cause concurrency exceptions because of the concurrent dictionary, but this scenario still might happen:
thread 1: trygetvalue returns false
thread 2: trygetvalue returns false
thread 1: create the new counter and insert into the dictionary
thread 2: creat a new counter and overwrite the previous one